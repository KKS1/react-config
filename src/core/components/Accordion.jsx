import React, { Component } from "react";
import Label from "./Label";

const AccordionSection = props => {
  const { option, id, index } = props;
  const headerId = id + "_header_" + index;
  const contentId = id + "_content_" + index;

  return (
    <div className="c-accordion__group">
      <a
        className="col-12 c-accordion__heading p-0 collapsed"
        data-toggle="collapse"
        id={headerId}
        data-parent={`#${id}`}
        data-target={`#${contentId}`}
        aria-controls={contentId}
        href="#"
        aria-expanded="false"
      >
        {option.title}
        <i className="c-accordion__icon" />
      </a>
      <div
        className="col-12 c-accordion__body collapse"
        data-parent={`#${id}`}
        role="region"
        aria-labelledby={headerId}
        id={contentId}
      >
        {option.content}
      </div>
    </div>
  );
};

export default class Accordion extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    const { label, options, id = _.uniqueId('accordion_'), ...restProps } = this.props;

    return (
      <React.Fragment>
        <Label {...this.props} />

        <div className="c-accordion" id={id}>
          {options &&
            options.map((option, index) => {
              const newProps = {
                ...restProps,
                id,
                option,
                index
              };
              return <AccordionSection key={index} {...newProps} />;
            })}
        </div>
      </React.Fragment>
    );
  }
}
