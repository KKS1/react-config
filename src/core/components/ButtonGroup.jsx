import React, { Component } from "react";
import classNames from "classnames";
import Label from "./Label";

const ButtonItem = props => {
  const { enumOptions, option, onChange, value, id, index } = props;
  const isActive = value === option.value;
  const btnClassName = classNames("btn", { active: isActive });

  return (
    <button
      type="button"
      className={btnClassName}
      onClick={event => onChange(option.value)}
    >
      {option.label}
    </button>
  );
};

export default class ButtonGroup extends Component {
  static defaultProps = {
    name: "btnGroup",
    onChange: () => {}
  };

  render() {
    const { name, value, id, options, label } = this.props;
    const { enumOptions } = this.props.options;

    return (
      <div className="c-label-field">
        <Label for={id} {...this.props} />
        <div
          className="btn-group"
          role="group"
          id={id}
          name={name}
          aria-label={label}
        >
          {enumOptions &&
            enumOptions.map((option, index) => {
              const newProps = { ...this.props, enumOptions, option, index };
              return <ButtonItem key={index} {...newProps} />;
            })}
        </div>
      </div>
    );
  }
}
