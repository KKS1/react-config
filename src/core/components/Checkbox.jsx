import React, { Component } from "react";

const Checkbox = props => {
  const { value, label, onChange, name, id } = props;
  return (
    <label className="c-checkbox">
      <input
        id={id}
        className="c-checkbox__input"
        name={name}
        type="checkbox"
        checked={typeof value === "undefined" ? false : value}
        value={value}
        onChange={event => props.onChange(event.target.checked)}
      />
      <i className="c-checkbox__icon" />
      <span className="c-checkbox__label">{label}</span>
    </label>
  );
};

export default Checkbox;
