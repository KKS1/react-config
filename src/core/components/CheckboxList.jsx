import React, { Component } from "react";
import Label from "./Label";

const selectValue = (value, selected = [], all) => {
  const at = all.indexOf(value);
  const updated = selected.slice(0, at).concat(value, selected.slice(at));
  return updated.sort((a, b) => all.indexOf(a) > all.indexOf(b));
};

const deselectValue = (value, selected) => {
  return selected.filter(v => v !== value);
};

const CheckboxItem = props => {
  const { enumOptions, option, onChange, value, id, index, label } = props;
  const checked = value ? value.indexOf(option.value) !== -1 : false;
  const classList = "js-multi-select__option c-multi-select__option c-checkbox";
  const optionId = `${id}_${index}`;
  const optionName = `${id}[0]`;

  const changeHandler = ({ event, option, value }) => {
    const all = enumOptions.map(({ value }) => value);
    const newValue = event.target.checked
      ? selectValue(option.value, value, all)
      : deselectValue(option.value, value);

    onChange(newValue);
  };

  return (
    <div className="col-12 col-md-10" role="listitem" htmlFor={optionId}>
      <label className={classList}>
        <input
          id={optionId}
          className="c-checkbox__input"
          name={optionName}
          type="checkbox"
          checked={checked}
          value={option.value}
          onChange={event => changeHandler({ event, option, value })}
        />
        <i className="c-checkbox__icon" />
        <span className="c-checkbox__label">{option.label}</span>
      </label>
    </div>
  );
};

export default class CheckboxList extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    const label = this.props.label
      ? this.props.label
      : this.props.schema && this.props.schema.title
      ? this.props.schema.title
      : "";
    const { enumOptions } = this.props.options;

    return (
      <React.Fragment>
        <Label {...this.props} />
        <fieldset
          id={this.props.id}
          className="js-multi-select c-multi-select row c-multi-select--condensed"
          role="list"
        >
          {enumOptions &&
            enumOptions.map((option, index) => {
              const newProps = {
                ...this.props,
                enumOptions,
                option,
                index,
                label
              };
              return <CheckboxItem key={index} {...newProps} />;
            })}
        </fieldset>
      </React.Fragment>
    );
  }
}
