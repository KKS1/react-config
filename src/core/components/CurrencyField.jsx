import React, { Component } from "react";
import MaskedInput from "./MaskedInput";
import Label from "./Label";

export default class CurrencyField extends Component {
  static defaultProps = {
    label: "Label",
    placeholder: "",
    name: "currencyFormField",
    onChange: () => {}
  };

  changeHandler = e => {
    this.props.onChange(e.target.value);
  };

  render() {
    const {
      label,
      placeholder,
      name,
      value,
      id,
      options,
      disabled
    } = this.props;
    const { maskconfig } = options || {};

    return (
      <div className="c-label-field">
        <Label {...this.props} />
        <div className="d-flex">
          <i className="far fa-dollar-sign c-label-field__icon c-label-field__icon--prepend" />
          <MaskedInput
            id={id}
            className="c-label-field__input c-label-field__input--icon-prefixed"
            type="text"
            inputMode="numeric"
            placeholder={placeholder}
            name={name}
            disabled={disabled}
            value={value == null ? 0 : value}
            maskconfig={maskconfig}
            onChange={this.changeHandler}
          />
        </div>
      </div>
    );
  }
}
