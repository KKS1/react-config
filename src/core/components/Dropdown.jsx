import React, { Component } from "react";
import styled from "styled-components";
import Label from "./Label";

const WrapperDiv = styled.div`
  padding: 0px;
`;

const HiddenLabel = styled.label`
  display: none;
`;

const OptionItem = ({ optionData }) => {
  const { value, label } = optionData;
  return <option value={value}>{label}</option>;
};

export default class Dropdown extends Component {
  changeHandler = e => {
    this.props.onChange(e.target.value);
  };

  render() {
    const { name, value, id, options, labelby, customClasses } = this.props;
    const { enumOptions } = options;

    return (
      <WrapperDiv className={`c-label-field ${customClasses}`}>
        {!labelby && <Label {...this.props} />}
        <div className="c-select">
          <select
            id={id}
            name={name}
            value={typeof value === "undefined" ? "" : value}
            className="c-select__menu"
            aria-labelledby={labelby}
            onChange={this.changeHandler}
          >
            {enumOptions.map((optionData, index) => (
              <OptionItem key={index} optionData={optionData} />
            ))}
          </select>
        </div>
      </WrapperDiv>
    );
  }
}
