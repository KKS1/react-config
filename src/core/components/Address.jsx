import React, { Component } from "react";
import styled from "styled-components";
import Label from "./Label";
import LocationAutoComplete from "./LocationAutoComplete";
import TextField from "./TextField";
import Dropdown from "./Dropdown";

const STATES = [
  "",
  "AK",
  "AL",
  "AZ",
  "AR",
  "CA",
  "CO",
  "CT",
  "DE",
  "DC",
  "FL",
  "GA",
  "HI",
  "ID",
  "IL",
  "IN",
  "IA",
  "KS",
  "KY",
  "LA",
  "ME",
  "MD",
  "MA",
  "MI",
  "MN",
  "MS",
  "MO",
  "MT",
  "NE",
  "NV",
  "NH",
  "NJ",
  "NM",
  "NY",
  "NC",
  "ND",
  "OH",
  "OK",
  "OR",
  "PA",
  "RI",
  "SC",
  "SD",
  "TN",
  "TX",
  "UT",
  "VT",
  "VA",
  "WA",
  "WV",
  "WI",
  "WY"
];

const dropdownOptions = {
  enumOptions: STATES.map(state => ({
    value: state,
    label: state ? state : "Please select one"
  }))
};

const zipTextFieldOptions = {
  inputProps: { className: "w-50" }
};

const addressProps = {
  ADDRESS_1: "address1",
  ADDRESS_2: "address2",
  CITY: "city",
  STATE: "state",
  ZIP: "zip"
};

const AddressContainer = styled.div`
  & > * {
    margin-bottom: 1rem;
  }
`;

export default class Address extends Component {
  static defaultProps = {
    label: "Address",
    placeholder: "",
    name: "addressWidget",
    onChange: () => {}
  };

  constructor(props) {
    super(props);
  }

  componentDidMount() {}

  propogateOnChange = (prop, value) => {
    this.props.onChange({ ...this.props.formData, [prop]: value });
  };

  changeHandler = (prop, val) => {
    this.propogateOnChange(prop, val);
  };

  placeChangeHandler = ({ formatted_address }) => {
    if (!formatted_address) {
      return;
    }

    const placesArr = formatted_address.split(", ");
    const [country, stateAndZip, city, ...rest] = placesArr.reverse();
    const address1 = rest && rest.join(", ");
    const [state, zip] = stateAndZip.split(" ");

    const updatedDetails = {
      address1,
      city,
      state,
      zip
    };

    Object.keys(updatedDetails).forEach(prop =>
      this.changeHandler(prop, updatedDetails[prop])
    );
  };

  render() {
    const { placeholder, name, value, formData, idSchema } = this.props;
    const { address1, address2, city, state, zip } = formData;
    const { $id: id } = idSchema;

    return (
      <AddressContainer>
        <LocationAutoComplete
          id={`${id}_${addressProps.ADDRESS_1}`}
          label="Address"
          value={address1}
          onChange={val => this.changeHandler(addressProps.ADDRESS_1, val)}
          onPlaceChange={this.placeChangeHandler}
        />

        <TextField
          id={`${id}_${addressProps.ADDRESS_2}`}
          label="Address Line 2"
          value={address2}
          onChange={val => this.changeHandler(addressProps.ADDRESS_2, val)}
        />

        <TextField
          id={`${id}_${addressProps.CITY}`}
          label="City"
          value={city}
          onChange={val => this.changeHandler(addressProps.CITY, val)}
        />

        <Dropdown
          id={`${id}_${addressProps.STATE}`}
          label="State"
          customClasses="w-50"
          options={dropdownOptions}
          value={state}
          onChange={val => this.changeHandler(addressProps.STATE, val)}
        />

        <TextField
          id={`${id}_${addressProps.ZIP}`}
          label="Zip"
          options={zipTextFieldOptions}
          value={zip}
          onChange={val => this.changeHandler(addressProps.ZIP, val)}
        />
      </AddressContainer>
    );
  }
}
