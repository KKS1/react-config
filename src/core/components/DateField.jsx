import React, { Component } from "react";
import Dropdown from "./Dropdown";
import Label from "./Label";

/** TODOS:
 * - use moment.js for date management and generation ( moment.months() / moment("2019-Feb", "YYYY-MMM").daysInMonth() )
 **/

const MONTH_LABELS = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December"
];

const getDefaultYearsRange = () => {
  const currentYear = new Date().getFullYear();
  return [currentYear - 50, currentYear];
};

const pad = (num, size) => {
  let s = String(num);
  while (s.length < size) {
    s = "0" + s;
  }
  return s;
};

const rangeOptions = ({ range, defaultOption, labelFn, sortDesc = false }) => {
  let options = [];
  const start = range[0];
  const stop = range[1];

  const optionLabelFn = labelFn || (val => val.toString());

  for (let i = start; i <= stop; i++) {
    options.push({ value: parseInt(i), label: optionLabelFn(i) });
  }

  options = sortDesc
    ? [...options, defaultOption].reverse()
    : [defaultOption, ...options];

  return options;
};

const months = rangeOptions({
  range: [1, 12],
  defaultOption: { value: 0, label: "Month" },
  labelFn: i => MONTH_LABELS[i - 1]
});

const days = rangeOptions({
  range: [1, 31],
  defaultOption: { value: 0, label: "Day" },
  labelFn: i => pad(i, 2)
});

const dropdownClasses = {
  month: "mr-3 w-50",
  day: "mr-3 w-25",
  year: "mr-3 w-25"
};

export default class DateField extends Component {
  static defaultProps = {
    formData: {}
  };

  constructor(props) {
    super(props);
  }

  propogateOnChange = (prop, value) => {
    this.props.onChange({ ...this.props.formData, [prop]: value });
  };

  changeHandler = (prop, val) => {
    this.propogateOnChange(prop, parseInt(val));
  };

  render() {
    const {
      schema,
      uiSchema = {},
      placeholder,
      name,
      formData,
      idSchema
    } = this.props;
    const { $id: id } = idSchema;
    const { title: label } = schema;
    const { "ui:options": options = {} } = uiSchema;
    const {
      yearsRange = getDefaultYearsRange(),
      hiddenFields = [],
      prefix = ""
    } = options;

    const dropdownData = {
      month: { enumOptions: months },
      day: { enumOptions: days },
      year: {
        enumOptions: rangeOptions({
          range: yearsRange,
          defaultOption: { value: 0, label: "Year" },
          sortDesc: true
        })
      }
    };

    return (
      <div className="c-label-field">
        <label className="control-label" id={id}>
          {label}
        </label>
        <div className="d-flex">
          {Object.keys(dropdownData).map((key, index) => {
            const dropdownId = id + "_" + key;
            const fieldProp = prefix + key;
            return (
              !hiddenFields.includes(key) && (
                <Dropdown
                  id={dropdownId}
                  key={index}
                  customClasses={dropdownClasses[key]}
                  options={dropdownData[key]}
                  value={formData[fieldProp]}
                  labelby={id}
                  onChange={val => this.changeHandler(fieldProp, val)}
                />
              )
            );
          })}
        </div>
      </div>
    );
  }
}
