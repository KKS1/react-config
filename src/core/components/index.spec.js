import Accordion from "./Accordion";
import Address from "./Address";
import ButtonGroup from "./ButtonGroup";
import Checkbox from "./Checkbox";
import CheckboxList from "./CheckboxList";
import CurrencyField from "./CurrencyField";
import DateField from "./DateField";
import Dropdown from "./Dropdown";
import Footer from "./Footer";
import Header from "./Header";
import Label from "./Label";
import LocationAutoComplete from "./LocationAutoComplete";
import MaskedInput from "./MaskedInput";
import Paragraph from "./Paragraph";
import PercentField from "./PercentField";
import RadioGroup from "./RadioGroup";
import TextField from "./TextField";
import Zillow from "./Zillow";

import Atomic from "./index";

describe("Atomic", () => {
  test("contains_appropriate_objects", () => {
    expect(Atomic).toMatchSnapshot();
  });
});
