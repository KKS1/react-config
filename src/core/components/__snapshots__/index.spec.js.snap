// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atomic contains_appropriate_objects 1`] = `
Object {
  "Accordion": [Function],
  "Address": [Function],
  "ButtonGroup": [Function],
  "Checkbox": [Function],
  "CheckboxList": [Function],
  "CurrencyField": [Function],
  "DateField": [Function],
  "Dropdown": [Function],
  "Footer": [Function],
  "Header": [Function],
  "Label": [Function],
  "LocationAutoComplete": [Function],
  "MaskedInput": [Function],
  "NumericField": [Function],
  "Paragraph": [Function],
  "PercentField": [Function],
  "RadioGroup": [Function],
  "TextField": [Function],
  "Zillow": [Function],
}
`;
